src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,461 @@
+            var startPoint = openingPoint;
+            var endPoint = closingPoint;
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Only format outside of the completed braces if they're on the same line for array/collection/object initializer expressions.
+            // Example:   `var x = new int[]{}`:
+            // Correct:   `var x = new int[] {}`
+            // Incorrect: `var x = new int[] { }`
+            // This is a heuristic to prevent brace completion from breaking user expectation/muscle memory in common scenarios.
+            // see bug Devdiv:823958
+            if (text.Lines.GetLineFromPosition(startPoint) == text.Lines.GetLineFromPosition(endPoint))
+            {
+                var startToken = root.FindToken(startPoint, findInsideTrivia: true);
+                if (startToken.IsKind(SyntaxKind.OpenBraceToken) &&
+                    (startToken.Parent?.IsInitializerForArrayOrCollectionCreationExpression() == true ||
+                     startToken.Parent is AnonymousObjectCreationExpressionSyntax))
+                {
+                    // Since the braces are next to each other the span to format is everything up to the opening brace start.
+                    endPoint = startToken.SpanStart;
+                }
+            }
+
+            var style = documentOptions.GetOption(FormattingOptions.SmartIndent);
+            if (style == FormattingOptions.IndentStyle.Smart)
+            {
+                // Set the formatting start point to be the beginning of the first word to the left 
+                // of the opening brace location.
+                // skip whitespace
+                while (startPoint >= 0 && char.IsWhiteSpace(text[startPoint]))
+                {
+                    startPoint--;
+                }
+
+                // skip tokens in the first word to the left.
+                startPoint--;
+                while (startPoint >= 0 && !char.IsWhiteSpace(text[startPoint]))
+                {
+                    startPoint--;
+                }
+            }
+
+            var spanToFormat = TextSpan.FromBounds(Math.Max(startPoint, 0), endPoint);
+            var rules = document.GetFormattingRules(braceFormattingIndentationRules, spanToFormat);
+            var result = Formatter.GetFormattingResult(root, SpecializedCollections.SingletonEnumerable(spanToFormat), document.Project.Solution.Workspace, documentOptions, rules, cancellationToken);
+            if (result == null)
+            {
+                return (ImmutableArray<TextChange>.Empty, closingPoint);
+            }
+
+            var newRoot = result.GetFormattedRoot(cancellationToken);
+            var newClosingPoint = newRoot.GetAnnotatedTokens(s_closingBraceSyntaxAnnotation).Single().SpanStart + 1;
+
+            var textChanges = result.GetTextChanges(cancellationToken).ToImmutableArray();
+            return (textChanges, newClosingPoint);
```

Comment:
main question: is this tested well?  i don't understand a lick of it.  but if we have good test coverage (and this is mostly moved code) then i'm ok with that.

Reply:
Yeah, we have quite a few unit tests and integration tests for all the brace completion scenarios, so I'm relatively confident in these changes